elfloader-tool/include/arch-arm/elfloader.h

```diff
@@ -9,9 +10,10 @@
 
 typedef void (*init_arm_kernel_t)(paddr_t ui_p_reg_start,
                                   paddr_t ui_p_reg_end,
-                                  uintptr_t pv_offset,
+                                  sword_t pv_offset,
```

Comment:
Right, I see the comment further down now.

Reply:
My understanding of the code is, that this is called "offset", but it seem this was designed to be a positive of negative offset. So there is no strict requirement the root task run at a virtual address that is less than the physical address - even if this is quite likely because user code is usually linked somewhere around the bottom of the address space. It seem a lot of code in influenced by the fact that this used to be a positive value everywhere. 
I was wondering if we might consider doing a bigger breaking change and pass `virt_start" instead of `pv_offset` here, so there is no need to ever think about signed values at inter-binary interfaces. That would put the problem into the binaries (loader, kernel) that process this, where they never really need to thing about this "delta" in the end, because the address spaces are distinct - except kernel's internal mapping concept with the kernel windows, but that's an internal thing to handle then.

Reply:
For me this is sufficient. `offset` also makes it clear that we're expecting the translation to be linear and additive.

As an aside: on the binary level, for the platforms we have, there is no difference between signed and unsigned values for most operations. The only things that are really influenced by signed/unsigned are comparison (there the sign really matters) and overflow (not allowed for `signed` in C). I.e. in the proofs, those operations are the only ones that actually need to know if the type was signed or unsigned. For most arithmetic operations, e.g. `+`, `-` etc the same binary code will produce correct results, no matter if you give it signed or unsigned input. I.e. `virt_address + phys_virt_offset = phys_address` will hold correctly even if the offset is negative, because the unsigned interpretation of a signed value produces the correct overflow.

Anyway, better to make it explicit when a potentially negative value is expected, instead of relying on bit magic that happens to be true for the platforms we use (it's not true for all number representations, of course).

Reply:
I think making this a signed word can make the situation also unclear.  The sign can't be used to represent the direction of the offset unless the type's width is increased to be able to hold a word sized offset and a direction.  So if you had an offset from 0x00000000 to 0xc0000000 in a 32bit architecture, this offset couldn't be stored in an sword_t without some sort of overflow in which case the sign is no longer representing the direction of the offset.

Reply:
Kent has a good point there. If you're starting at 0 and want to get to `0xc0000000` you'd have to use a negative offset, which would look quite strange. I think overflow-wise that would be Ok, but the fact that you have to think about that deeply is not good.

It might make more sense to add a comment that the offset is allowed to be negative in the sense that it is Ok for it to wrap around. unsigned overflow is allowed and well defined, so there are no problems on that side. The type then doesn't indicate the intention that the offset can go in either direction, but I think that's more acceptable than introducing worries about overflow, esp when we don't necessarily see all of the code that computes that offset, which may run into undefined behaviour if it is not done right.

Making a larger interface change (`virt_start`) will cause additional work for everyone who has their own boot loader, so I'm still not that keen on that option.